pbmtools/jbigtopbm.1
.TH JBIGTOPBM 1 "1995-05-28"
.SH NAME
jbigtopbm \- JBIG to portable bitmap file converter
.SH SYNOPSIS
.B jbigtopbm
[
.I options
]
[
.I input-file
| \-  [
.I output-file
]]
.br
.SH DESCRIPTION
Reads in a 
.I JBIG
bi-level image entity (BIE) from a file or standard
input, decompresses it and outputs a portable bitmap (PBM) file.

.I JBIG
is a highly effective lossless compression algorithm for
bi-level images (one bit per pixel) which is particularly suitable
for scanned document pages.

A
.I JBIG
encoded image can be stored in several resolutions in one or several BIEs.
All resolution layers except the lowest one are stored efficently as
differences to the next lower resolution layer.
Options
.BI -x
and
.BI -y
can be used to stop the decompression at a specified maximal output
image size. The input file can consist several concatenated BIEs which
contain different increasing resolution layers of the same image.

If more than one bit per pixel is stored in the JBIG file, then only
the first plane will be extracted and a warning message will be printed.
.SH OPTIONS
.TP 14
.BI \-
A single hyphen instead of an input file name will cause 
.I jbigtopbm
to read the data from standard input instead from a file.
.TP
.BI \-x " number"
Decode only up to the largest resolution layer which is still not
more than
.I number
pixels wide. If no such resolution layer exists, then use the smallest
one available.
.TP
.BI \-y " number"
Decode only up to the largest resolution layer which is still not
more than
.I number
pixels high. If no such resolution layer exists, then use the smallest
one available. Options
.BI \-x
and
.BI \-y
can also be used together in which case the largest layer that satisfies
both limits is selected.
.SH STANDARDS
This program implements the
.I JBIG
image coding algorithm as specified in ISO/IEC 11544:1993 and
ITU-T T.82(1993).
.SH AUTHOR
The
.I jbigtopbm 
program is part of the 
.I JBIG-KIT
package, which has beed developed by Markus Kuhn.
The most recent version of this
portable
.I JBIG
library and tools set is freely available on the Internet
from anonymous ftp server ftp.uni-erlangen.de in directory pub/doc/ISO/JBIG/.
Bug reports should be sent to <mskuhn@cip.informatik.uni-erlangen.de>. 
.SH SEE ALSO
pbm(5), pbmtojbig(1)


CHANGES

JBIG-KIT revision history
-------------------------


Changes in version 2.1 (2014-04-08)

This is a security-critical bug-fix release that remains API and ABI
backwards compatible to version 2.0. Users who process BIE data from
untrusted sources should upgrade.

  - fixed a buffer-overflow vulnerability in the jbig.c decoder,
    reported by Florian Weimer (Red Hat): CVE-2013-6369

  - fixed ability of corrupted input data to force jbig85.c decoder
    into an end-less loop

  - fixed a bug in the processing of private deterministic-prediction
    tables (DPPRIV=1) in jbig.c decoder

  - fixed integer-type mismatches in printf arguments on 64-bit systems

  - fuzz-testing script added


Changes in version 2.0 (2008-08-30)

Main new features of release 2.0:

  - This release adds a separate lightweight "T.85" version of the
    encoder and decoder library (jbig85.c, jbig85.h). This new T.85
    library is optimized for embedded low-memory applications, such as
    printers and fax machines. It implements only the subset of the
    JBIG standard that is required by the ITU-T T.85 fax profile,
    namely only a single bit plane (P = 1) and no differential layers
    (DL = D = HITOLO = SEQ = ILEAVE = SMID = TPDON = DPON = DPON =
    DPLAST = 0) and some other restrictions (MY = 0, MX < 128).

    The T.85 library requires only three pixel rows of the
    uncompressed image to reside in memory at any time. This 3-line
    buffer has to be allocated by the calling application, therefore
    the T.85 library will not call any heap management functions such
    as malloc() or free() itself. It can fully deal with endless-paper
    fax, where the height of the encoded image is not known initially.

  - There are also two new example applications pbmtojbg85 and
    jbgtopbm85 included that provide command-line access to the new
    T.85 library.

API changes in 2.0 to the original full jbig.c library (compared to 1.6):

  - jbg_dec_getwidth(), jbg_dec_getheight(), jbg_dec_getsize(), and
    jbg_dec_getsize_merged() return now an unsigned long value
    (was: long), and return 0 (was: -1) if no image is available yet

  - jbg_dec_in() return values changed; the least significant bits of
    the error codes now provide more information about what caused the
    error, beyond what the more general error text messages returned by
    jbg_strerror() reveal

  - non-English translations of error messages are now provided
    in a separate PO file (e.g., for use with GNU gettext),
    jbg_strerror() no longer has a language attribute and the
    JBG_EN, JBG_DE_8859_1, JBG_DE_UTF_8 constants no longer exist

Other changes in 2.0 to the original full jbig.c library:

  - the arithmetic codec routines are now located in a separate source
    file (they are now used by both the full library jbig.c and the new
    T.85 library jbig85.c, and might also be of use to other applications)

  - arithmetic decoder behaviour slightly changed to enable the
    lookahead needed to handle T.85-style NEWLEN after the final
    stripe

  - added encoder support for adding comment marker segment

  - added encoder option for using SDRST marker instead of SDNORM
    (only useful to make encoder able to generate a wider range
    of test data)

  - macro JBG_LICENCE can be changed from the default value "GPL" to a
    licence agreement reference code to if the library is used under a
    commercial licence, to clarify under which exact licence the
    library is used in a particular application

Changes in 2.0 to the command-line wrappers for jbig.c:

  - added pbmtojbg options -C and -r, which provide command-line
    access to the previous two extensions

  - slightly improved diagnostic output of jbgtopbm option -d

  - pbmtojbg has a new option -f as a shortcut for all the options
    needed to make the output comply to the basic ITU-T T.85 fax
    profile

  - jbgtopbm now outputs image dimensions in header padded to 10
    digits fixed width, for consistency with jbgtopbm85
    to allow for later in-place update of height due to NEWLEN


Changes in version 1.6 (2004-06-11)

  - various small changes to reduce the risk of 32-bit unsigned
    integer overflows when dealing with extremely large images

  - robuster treatment of L0 = 0xffffffff

  - minor API modification in jbg_enc_options(): parameter l0 changed
    from type long to unsigned long; previous value now remains
    unchanged when l0 == 0 (was: l0 < 0)

  - lots of type casts added such that the C source code is now
    also compilable as C++


Changes in version 1.5 (2003-06-11)

  - fixed two minor memory leaks (special thanks to Davide Pizzolato
    <ing.davide.pizzolato@libero.it> for locating one of these)

  - jbgtopbm does not attempt any more to parse multiple concatenated
    BIEs (options -m must be used now to enable this feature explicitly),
    in order to handle BIEs with data after the last expected SDE gracefully

  - various extensions to improve support of JBIG fax applications
    (ITU-T T.85 profile):

      o support for maximum adaptive template pixel offset increased
        to MX=127 in both encoder and decoder

      o encoder now has a hook for producing BIEs with a NEWLEN marker
        segment and VLENGTH=1, in order to assist in testing decoders
        for T.85 conformance (see also new pbmtojbg option -Y)

      o a new function jbg_newlen() can be used to scan an
        entire in-memory BIE for NEWLEN marker segments and update the
        YD value in the BIE header, which can be applied to BIE data
        before passing it to the decoder for T.85 compliance
        (Background: the incremental design of the JBIG-KIT decoder
        makes it infeasible to look ahead for NEWLEN marker segments
        that occur after the SDE with the last image line, therefore
        this second pass is necessary to handle the old-style fax
        applications permitted by T.85 in which the low-RAM encoder

pbmtools/jbigtopbm.c
/*
 *  jbigtopbm - JBIG to Portable Bitmap converter
 *
 *  Markus Kuhn -- mskuhn@cip.informatik.uni-erlangen.de
 *
 *  $Id: jbgtopbm.c,v 1.2 1995-06-08 16:44:30 mskuhn Exp $
 */

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include "jbig.h"

const char usage_msg[] = "JBIGtoPBM converter 0.6 -- "
"reads bi-level image entity (BIE) as input file\n\n"
"usage: %s [<options>] [<input-file> | -  [<output-file>]]\n\n"
"options:\n\n"
"  -x number\tif possible decode only up to a resolution layer not\n"
"\t\t\twider than the given number of pixels\n"
"  -y number\tif possible decode only up to a resolution layer not\n"
"\t\t\thigher than the given number of pixels\n\n";

char *progname;                  /* global pointer to argv[0] */


/*
 * Print usage message and abort
 */
void usage(void)
{
  fprintf(stderr, usage_msg, progname);
  exit(1);
}


int main (int argc, char **argv)
{
  FILE *fin = stdin, *fout = stdout;
  char *fnin = "<stdin>", *fnout = "<stdout>";
  int i, j, result;
  int all_args = 0, files = 0;
  struct jbg_dec_state s;
#define BUFSIZE 8192
  char buffer[BUFSIZE];
  unsigned char *p;
  size_t len, cnt;
  unsigned long xmax = 4294967295U, ymax = 4294967295U;

  /* parse command line arguments */
  progname = argv[0];
  for (i = 1; i < argc; i++) {
    if (!all_args && argv[i][0] == '-')
      if (argv[i][1] == '\0' && files == 0)
        ++files;
      else
        for (j = 1; j < 10000 && argv[i][j]; j++)
          switch(tolower(argv[i][j])) {
          case '-' :
            all_args = 1;
            break;
          case 'x':
            if (++i >= argc) usage();
            j = 10000;
            xmax = atol(argv[i]);
            break;
          case 'y':
            if (++i >= argc) usage();
            j = 10000;
            ymax = atol(argv[i]);
            break;
          default:
            usage();
          }
    else
      switch (files++) {
      case 0:
        if (argv[i][0] != '-' || argv[i][1] != '\0') {
          fnin = argv[i];
          fin = fopen(fnin, "rb");
          if (!fin) {
            fprintf(stderr, "Can't open input file '%s", fnin);
            perror("'");
            exit(1);
          }
        }
        break;
      case 1:
        fnout = argv[i];
        fout = fopen(fnout, "wb");
        if (!fout) {
          fprintf(stderr, "Can't open input file '%s", fnout);
          perror("'");
          exit(1);
        }
        break;
      default:
        usage();
      }
  }

  /* send input file to decoder */
  jbg_dec_init(&s);
  jbg_dec_maxsize(&s, xmax, ymax);
  result = JBG_EAGAIN;
  do {
    len = fread(buffer, 1, BUFSIZE, fin);
    if (!len) break;
    cnt = 0;
    p = (unsigned char *) buffer;
    while (len > 0 && (result == JBG_EAGAIN || result == JBG_EOK)) {
      result = jbg_dec_in(&s, p, len, &cnt);
      p += cnt;
      len -= cnt;
    }
  } while (result == JBG_EAGAIN || result == JBG_EOK);
  if (ferror(fin)) {
    fprintf(stderr, "Problem while reading input file '%s", fnin);
    perror("'");
    exit(1);
  }
  if (result != JBG_EOK && result != JBG_EOK_INTR) {
    fprintf(stderr, "Problem with input file '%s': %s\n",
	    fnin, jbg_strerror(result, JBG_EN));
    exit(1);
  }
  if (jbg_dec_getplanes(&s) > 1)
    fprintf(stderr, "Warning: storing only first of %d bit planes, ignoring "
	    "remaining %d planes.\n",
	    jbg_dec_getplanes(&s), jbg_dec_getplanes(&s) - 1);
  
  /* write PBM output file */
  fprintf(fout, "P4\n%ld %ld\n", jbg_dec_getwidth(&s), jbg_dec_gethight(&s));
  fwrite(jbg_dec_getimage(&s, 0), 1,
	 jbg_dec_getsize(&s), fout);

  /* check for file errors and close fout */
  if (ferror(fout) || fclose(fout)) {
    fprintf(stderr, "Problem while writing output file '%s", fnout);
    perror("'");
    exit(1);
  }

  jbg_dec_free(&s);
  
  return 0;
}
